vms/platformvm/genesis/genesis_builder.go

```diff
@@ -0,0 +1,329 @@
+	errValidatorAlreadyExited = errors.New("validator would have already unstaked")
+	errStakeOverflow          = errors.New("validator stake exceeds limit")
+
+	_ utils.Sortable[Allocation] = Allocation{}
+)
+
+// Allocation is a UTXO on the Platform Chain that exists at the chain's genesis
+type Allocation struct {
+	Locktime uint64
+	Amount   uint64
+	Address  string
+	Message  string
+}
+
+// Compare compares two allocations
+func (utxo Allocation) Compare(other Allocation) int {
+	if locktimeCmp := cmp.Compare(utxo.Locktime, other.Locktime); locktimeCmp != 0 {
+		return locktimeCmp
+	}
+	if amountCmp := cmp.Compare(utxo.Amount, other.Amount); amountCmp != 0 {
+		return amountCmp
+	}
+
+	utxoAddr, err := bech32ToID(utxo.Address)
+	if err != nil {
+		return 0
+	}
+
+	otherAddr, err := bech32ToID(other.Address)
+	if err != nil {
+		return 0
+	}
+
+	return utxoAddr.Compare(otherAddr)
+}
+
+// GenesisValidator represents a validator at genesis
+type GenesisValidator struct {
+	TxID      ids.ID
+	StartTime uint64
+	EndTime   uint64
+	Weight    uint64
+	NodeID    ids.NodeID
+}
+
+// Owner is the repr. of a reward owner at genesis
+type Owner struct {
+	Locktime  uint64
+	Threshold uint32
+	Addresses []string
+}
+
+// GenesisPermissionlessValidator represents a permissionless validator at genesis
+type GenesisPermissionlessValidator struct {
+	GenesisValidator
+	RewardOwner        *Owner
+	DelegationFee      float32
+	ExactDelegationFee uint32
+	Staked             []Allocation
+	Signer             *signer.ProofOfPossession
+}
+
+// Chain defines a chain that exists at the network's genesis
+// [GenesisData] is the initial state of the chain.
+// [VMID] is the ID of the VM this chain runs.
+// [FxIDs] are the IDs of the Fxs the chain supports.
+// [Name] is a human-readable, non-unique name for the chain.
+// [SubnetID] is the ID of the subnet that validates the chain
+type Chain struct {
+	GenesisData string
+	VMID        ids.ID
+	FxIDs       []ids.ID
+	Name        string
+	SubnetID    ids.ID
+}
+
+// bech32ToID takes bech32 address and produces a shortID
+func bech32ToID(addrStr string) (ids.ShortID, error) {
+	_, addrBytes, err := address.ParseBech32(addrStr)
+	if err != nil {
+		return ids.ShortID{}, err
+	}
+	return ids.ToShortID(addrBytes)
+}
+
+// BuildGenesis builds the genesis state of the P-Chain (and thereby the Avalanche network.)
+// [avaxAssetID] is the ID of the AVAX asset
+// [networkID] is the ID of the network
+// [allocations] are the UTXOs on the Platform Chain that exist at genesis.
+// [validators] are the validators of the primary network at genesis.
+// [chains] are the chains that exist at genesis.
+// [time] is the Platform Chain's time at network genesis.
+// [initialSupply] is the initial supply of the AVAX asset.
+// [message] is the message to be sent to the genesis UTXOs.
+// [encoding] is the encoding of the message & chains genesis data.
+func BuildGenesis(
```

Comment:
nit: rename to `genesis.New` (just name all constructors `New*` to avoid confusion)